staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go

```diff
@@ -0,0 +1,172 @@
+					FSType:           volume.VsphereVolume.FSType,
+					VolumeAttributes: make(map[string]string),
+				},
+			},
+			AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+		},
+	}
+	if volume.VsphereVolume.StoragePolicyName != "" {
+		pv.Spec.CSI.VolumeAttributes[paramStoragePolicyName] = pv.Spec.VsphereVolume.StoragePolicyName
+	}
+	return pv, nil
+}
+
+// TranslateInTreePVToCSI takes a PV with VsphereVolume set from in-tree
+// and converts the VsphereVolume source to a CSIPersistentVolumeSource
+func (t *vSphereCSITranslator) TranslateInTreePVToCSI(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	if pv == nil || pv.Spec.VsphereVolume == nil {
+		return nil, fmt.Errorf("pv is nil or VsphereVolume not defined on pv")
+	}
+	csiSource := &v1.CSIPersistentVolumeSource{
+		Driver:           VSphereDriverName,
+		VolumeHandle:     pv.Spec.VsphereVolume.VolumePath,
+		FSType:           pv.Spec.VsphereVolume.FSType,
+		VolumeAttributes: make(map[string]string),
+	}
+	if pv.Spec.VsphereVolume.StoragePolicyName != "" {
+		csiSource.VolumeAttributes[paramStoragePolicyName] = pv.Spec.VsphereVolume.StoragePolicyName
+	}
+	pv.Spec.VsphereVolume = nil
+	pv.Spec.CSI = csiSource
+	return pv, nil
+}
+
+// TranslateCSIPVToInTree takes a PV with CSIPersistentVolumeSource set and
+// translates the vSphere CSI source to a vSphereVolume source.
+func (t *vSphereCSITranslator) TranslateCSIPVToInTree(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	if pv == nil || pv.Spec.CSI == nil {
+		return nil, fmt.Errorf("pv is nil or CSI source not defined on pv")
+	}
+	csiSource := pv.Spec.CSI
+	vsphereVirtualDiskVolumeSource := &v1.VsphereVirtualDiskVolumeSource{
+		FSType: csiSource.FSType,
+	}
+	volumeFilePath, ok := csiSource.VolumeAttributes[paramVolumeFilePathAttribute]
+	if ok {
+		vsphereVirtualDiskVolumeSource.VolumePath = volumeFilePath
+	}
+	pv.Spec.CSI = nil
+	pv.Spec.VsphereVolume = vsphereVirtualDiskVolumeSource
+	return pv, nil
+}
+
+// CanSupport tests whether the plugin supports a given persistent volume
+// specification from the API.  The spec pointer should be considered
+// const.
+func (t *vSphereCSITranslator) CanSupport(pv *v1.PersistentVolume) bool {
+	return pv != nil && pv.Spec.VsphereVolume != nil
+}
+
+// CanSupportInline tests whether the plugin supports a given inline volume
+// specification from the API.  The spec pointer should be considered
+// const.
+func (t *vSphereCSITranslator) CanSupportInline(volume *v1.Volume) bool {
+	return volume != nil && volume.VsphereVolume != nil
+}
+
+// GetInTreePluginName returns the name of the in tree plugin driver
+func (t *vSphereCSITranslator) GetInTreePluginName() string {
+	return VSphereInTreePluginName
+}
+
+// GetCSIPluginName returns the name of the CSI plugin
+func (t *vSphereCSITranslator) GetCSIPluginName() string {
+	return VSphereDriverName
+}
+
+func (t *vSphereCSITranslator) RepairVolumeHandle(volumeHandle, nodeID string) (string, error) {
```

Comment:
What does this function do?  Where is it called?

Reply:
RepairVolumeHandle is needed in VerifyVolumesAttached on the external attacher when we need to do strict volume handle matching to check VolumeAttachment attached status. vSphere CSI driver will be capable of understanding the in-tree VMDK volume path and CNS volume ID, so we do not need to repair or patch volume before we can verify whether that volume is attached.

Reply:
Can you add your explanation as comments for this function?

Reply:
done.